hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestScannerLeaseCount.java

```diff
@@ -0,0 +1,222 @@
+  private static Connection CONN;
+  private static Table TABLE;
+
+  private static SingleProcessHBaseCluster cluster;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    StartTestingClusterOption option =
+      StartTestingClusterOption.builder().rsClass(MockedQuotaManagerRegionServer.class).build();
+    cluster = UTIL.startMiniCluster(option);
+    UTIL.getAdmin().createTable(TableDescriptorBuilder.newBuilder(TABLE_NAME)
+      .setColumnFamily(ColumnFamilyDescriptorBuilder.of(FAM)).build());
+    Configuration conf = new Configuration(UTIL.getConfiguration());
+    CONN = ConnectionFactory.createConnection(conf);
+    TABLE = CONN.getTable(TABLE_NAME);
+    UTIL.loadTable(TABLE, FAM);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    try {
+      TABLE.close();
+    } catch (Exception ignore) {
+    }
+    try {
+      CONN.close();
+    } catch (Exception ignore) {
+    }
+    UTIL.shutdownMiniCluster();
+  }
+
+  @Before
+  public void before() {
+    SHOULD_THROW = false;
+  }
+
+  @Test
+  public void itIncreasesScannerCount() throws Exception {
+    long beforeScan = countLeases();
+    ResultScanner scanner = TABLE.getScanner(new Scan());
+    Thread.sleep(1000);
+    scanner.close();
+    long afterScan = countLeases();
+    assertTrue(beforeScan < afterScan);
+  }
+
+  @Test
+  public void itDoesNotIncreaseScannerLeaseCount() throws Exception {
+    SHOULD_THROW = true;
+    long beforeScan = countLeases();
+    ResultScanner scanner = TABLE.getScanner(new Scan());
+    Thread.sleep(1000);
+    scanner.close();
+    long afterScan = countLeases();
+    assertSame(beforeScan, afterScan);
+  }
+
+  public static final class MockedQuotaManagerRegionServer
+    extends SingleProcessHBaseCluster.MiniHBaseClusterRegionServer {
+    private final MockedRpcQuotaManager rpcQuotaManager;
+    private final LeaseManager leaseManager;
+
+    public MockedQuotaManagerRegionServer(Configuration conf)
+      throws IOException, InterruptedException {
+      super(conf);
+      this.rpcQuotaManager = new MockedRpcQuotaManager(this);
+      this.leaseManager =
+        new MockedLeaseManager(conf.getInt(HConstants.THREAD_WAKE_FREQUENCY, 10 * 1000));
+    }
+
+    @Override
+    public RegionServerRpcQuotaManager getRegionServerRpcQuotaManager() {
+      return rpcQuotaManager;
+    }
+
+    @Override
+    public LeaseManager getLeaseManager() {
+      return leaseManager;
+    }
+  }
+
+  public static final class MockedLeaseManager extends LeaseManager {
+    private final AtomicLong leasesCreated = new AtomicLong(0);
+
+    public MockedLeaseManager(int leaseCheckFrequency) {
+      super(leaseCheckFrequency);
+    }
+
+    public long getLeasesCreated() {
+      return leasesCreated.get();
+    }
+
+    @Override
+    public void addLease(Lease lease) throws LeaseStillHeldException {
+      super.addLease(lease);
+      leasesCreated.incrementAndGet();
+    }
+  }
+
+  private long countLeases() {
```

Comment:
This is really coarse-grained. Even limiting the check to a specific region server would prove a more reliable test. Extracting the scannerId and looking for it specifically will be the most reliable way to test here. I don't know if we have reasonable utilities to expose that information on the client through.

Reply:
Updated the test to be less coarse-grained